src/Temporalio/Bridge/ByteArrayRef.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Google.Protobuf;
+
+namespace Temporalio.Bridge
+{
+    internal class ByteArrayRef
+    {
+        public static readonly ByteArrayRef Empty = new(new byte[0]);
+        internal static readonly UTF8Encoding StrictUTF8 = new(false, true);
+
+        public static ByteArrayRef FromUTF8(string s)
+        {
+            if (s.Length == 0)
+            {
+                return Empty;
+            }
+            return new ByteArrayRef(StrictUTF8.GetBytes(s));
+        }
+
+        public static ByteArrayRef FromProto(IMessage p)
+        {
+            return new ByteArrayRef(p.ToByteArray());
+        }
+
+        public static ByteArrayRef FromMetadata(IEnumerable<KeyValuePair<string, string>> metadata)
```

Comment:
> Extra simple: Just use a proto rather than this custom text based format

It's not necessarily extra simple to maintain lang-specific protos for crossing the bridge. Note we use structs in many places for Rust and C# that we could just replace with protos but we choose not to.

> Slightly less wasteful, but more annoying to build/free: Make a specific byte ref type in C like this:

Yes, I could make a linked list of pairs. And if I need a `map<string, string>` that is not gRPC metadata I almost surely will. But this is specifically for metadata which has a nice no-newline property.

> Lets you get rid of the special newline handling.

But this is basically how HTTP headers and gRPC metadata are handled on the wire (though they use `:` for delimiting between key and value while I use newline, but they still use newline for pair delimiting).

> In general, I think it probably makes sense to represent these different "kinds" of byte arrays as different C structs

Usually agree, but after I typed it out, I discovered metadata was the only use for this and that newlines worked well as delimiters just like they do for HTTP headers. If we need general purpose string maps, I will definitely go make them. I only use newlines here because I can, not because I'm so lazy I impose unreasonable burdens on the caller.

Reply:
Even if it's the only use, I think the separated types add a lot of clarity while reading. They don't need to actually *do* anything, per se, they just make what each field is being used for more clear to a reader and future-proof the code a bit at a very low cost. I really think it's worth it.

We can keep the newline parsing if you don't think the other change is worth it, but I'd really advocate for the different types.

Reply:
To confirm, you're saying a struct with a single byte array ref in it instead of just the byte array ref directly? Byte array refs are the only way I have to pass byte arrays (and therefore strings) from C# to Rust. I'd hate to add a new byte array struct, but I can wrap it if needed.

Reply:
Wrapping in C would be annoying because of the extra pointer. I think, if you're gonna keep metadata the way it is, I'd just make type aliases in C, and then little wrapper structs in Rust that use https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent

This way you get some type safety on the Rust side in terms of using the right "type" of ref in each place, and it doesn't really cost you anything. They can implement `Deref` into a `ByteArrayRef` to avoid annoying `.0` stuff.

Alternatively, if that's too annoying, just type alias in Rust too, which at least signals some intent even if it doesn't provide safety.

Reply:
Yeah I mean that's pretty much what I was suggesting, it just happens to be a bit cumbersome to do in C in a meaningfully type safe way.

The aliases at least signal intent nicely though, and cost nothing. I'd do it.

Reply:
Added alias